File: bench/08_word_count/bench.py
"""Benchmark 8: Word frequency count (300,000 hash operations).

The timed loop runs `--n` ops, each on an input that differs by the repetition number in
VALUE, never in the amount of work, and folds every result into `sink`.  The `hash`
column is one canonical op's result; every lane must print the same one.
"""
import sys
import time

WORDS = ["the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog",
         "the", "fox", "and", "the", "dog", "are", "friends", "the", "end"]


def tally(ops, salt):
    freq = {}
    for i in range(ops):
        w = WORDS[(i + salt) % 18]
        freq[w] = freq.get(w, 0) + 1
    return freq["the"] * 100 + freq["end"]

def arg_n(dflt):
    """`--n N`: how many ops the timed loop runs (the harness calibrates it per lane)."""
    n = dflt
    for i, a in enumerate(sys.argv):
        if a == "--n" and i + 1 < len(sys.argv):
            n = int(sys.argv[i + 1])
    return max(n, 2)


def row(name, iters, us, items, result, sink):
    """One measured routine, in the row format every lane prints (bench/README.md)."""
    print("routine\titers\tus\tns_op\tpx\tns_px\thash")
    print(f"{name}\t{iters}\t{us}\t{us * 1000 // iters}\t{items}\t{us * 1000 / (iters * items):.3f}\t{result:x}")
    print(f"time: {us // 1000}ms sink={sink}")


def timed(n, op):
    t0 = time.perf_counter_ns()
    sink = 0
    for r in range(n):
        sink += op(r)
    return (time.perf_counter_ns() - t0) // 1000, sink

if __name__ == "__main__":
    n = arg_n(2)
    ops = 300_000
    us, sink = timed(n, lambda r: tally(ops, r & 1))
    row("word_count", n, us, ops, tally(ops, 0), sink)

File: bench/08_word_count/bench.rs
// Benchmark 8: Word frequency count (300,000 hash operations)
//
// One OP is a fresh table filled from 300,000 words of an 18-word cycle, entered at `salt`.  The timed loop runs `--n` ops, each on an input that differs by
// the repetition number — in VALUE, never in the amount of work — so no optimiser can
// hoist or collapse the kernel across repetitions, and folds every result into `sink`.
// The `hash` column is one canonical op's result: every lane must print the same one,
// or the lanes are not computing the same thing and their times do not compare.
use std::collections::HashMap;
use std::hint::black_box;
use std::time::Instant;

const WORDS: [&str; 18] = ["the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog",
                           "the", "fox", "and", "the", "dog", "are", "friends", "the", "end"];

fn tally(ops: i64, salt: i64) -> i64 {
    let mut freq: HashMap<&str, i64> = HashMap::new();
    for i in 0..ops {
        *freq.entry(WORDS[((i + salt) % 18) as usize]).or_insert(0) += 1;
    }
    freq["the"] * 100 + freq["end"]
}

/// `--n N`: how many ops the timed loop runs (the harness calibrates it per lane).
fn arg_n(dflt: i64) -> i64 {
    let args: Vec<String> = std::env::args().collect();
    let mut n = dflt;
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--n" && i + 1 < args.len() {
            n = args[i + 1].parse().unwrap_or(dflt);
            i += 1;
        }
        i += 1;
    }
    n.max(2)
}

/// One measured routine, in the row format every lane prints (bench/README.md).
fn row(name: &str, iters: i64, us: i64, items: i64, result: i64, sink: i64) {
    let ns = us * 1000 / iters;
    let per = (us * 1000) as f64 / (iters * items) as f64;
    println!("routine\titers\tus\tns_op\tpx\tns_px\thash");
    println!("{name}\t{iters}\t{us}\t{ns}\t{items}\t{per:.3}\t{result:x}");
    println!("time: {}ms sink={}", us / 1000, black_box(sink));
}

fn main() {
    let n = arg_n(10);
    let ops: i64 = 300_000;
    let t0 = Instant::now();
    let mut sink: i64 = 0;
    for r in 0..n {
        sink = sink.wrapping_add(tally(black_box(ops), black_box(r & 1)));
    }
    let us = t0.elapsed().as_micros() as i64;
    row("word_count", n, us, ops, tally(ops, 0), sink);
}
